src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <exception>
+#include <map>
+#include <set>
+#include <pthread.h>
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+#include <cudf_jni_apis.hpp>
+
+namespace {
+
+constexpr char const *RMM_EXCEPTION_CLASS = "ai/rapids/cudf/RmmException";
+
+class thread_state {
+public:
+    bool retry_oom_injected = false;
+    bool split_and_retry_oom_injected = false;
+    long task_id = -1;
+};
+
+template<typename T>
+class rollback {
+public:
+    T on_error;
+
+    rollback(T & on_error): on_error(on_error) {}
+
+    ~rollback() {
+      if (std::uncaught_exceptions() > 0) {
+        on_error();
+      }
+    }
+};
+
+class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
+public:
+  spark_resource_adaptor(JNIEnv *env, rmm::mr::device_memory_resource *mr)
+      : resource{mr} {
+    if (env->GetJavaVM(&jvm) < 0) {
+      throw std::runtime_error("GetJavaVM failed");
+    }
+  }
+
+  rmm::mr::device_memory_resource *get_wrapped_resource() { return resource; }
+
+  bool supports_get_mem_info() const noexcept override { return resource->supports_get_mem_info(); }
+
+  bool supports_streams() const noexcept override { return resource->supports_streams(); }
+
+  void associate_thread_with_task(long thread_id, long task_id) {
+    std::scoped_lock lock(state_mutex);
+    if (threads.find(thread_id) != threads.end()) {
+      throw std::invalid_argument("a thread can only be added if it is in the unknown state");
+    }
+    threads.insert({thread_id, {false, false, task_id}});
+    auto recover = [this, thread_id]() {
+      threads.erase(thread_id);
+    };
+    {
+      rollback rb(recover);
+
+      auto tasks_it = task_to_threads.find(task_id);
+      if (tasks_it == task_to_threads.end()) {
+        task_to_threads.insert({task_id, {thread_id}});
+      } else {
+        tasks_it->second.insert(thread_id);
+      }
+    }
+  }
+
+  void remove_thread_association(long thread_id) {
+    std::scoped_lock lock(state_mutex);
+    auto threads_at = threads.find(thread_id);
```

Comment:
if we are removing a `thread_id` from `thread`, and the thread wasn't there before, is that bad?

Reply:
It is not what we expect, but is it bad enough to throw an exception over?  Probably a warning, but logging from c++ is hard, so I just ignore it.

Reply:
Probably not worth that. I am guessing the calling code could call removeThreadAssociation many times and that's considered safe/ok.